python/pyspark/testing/utils.py

```diff
@@ -209,3 +234,79 @@ def check_error(
         self.assertEqual(
             expected, actual, f"Expected message parameters was '{expected}', got '{actual}'"
         )
+
+
+def assertDataFrameEqual(df: DataFrame, expected: DataFrame, ignore_row_order: bool = True):
+    if df is None and expected is None:
+        return True
+    elif df is None or expected is None:
+        return False
+
+    def compare_rows(r1: Row, r2: Row):
+        def compare_vals(val1, val2):
+            if isinstance(val1, list) and isinstance(val2, list):
+                return all([compare_vals(x, y) for x, y in list(zip(val1, val2))])
+            elif isinstance(val1, dict) and isinstance(val2, dict):
+                return all([compare_vals(x, y) for x, y in list(zip(val1.values(), val2.values()))])
```

Comment:
The keys should also be tested?
Also, `val1.values()` doesn't guarantee the order of the values. We should iterate by their keys:

```py
return val1.keys() == val2.keys() and all([compare_vals(val1[k], val2[k]) for k in val1.keys()])
```